rfd/0057-automatic-aws-server-discovery.md

```diff
@@ -0,0 +1,347 @@
+---
+authors: Alex McGrath (alex.mcgrath@goteleport.com)
+state: draft
+---
+
+# RFD 57 - Automatic discovery and enrollment of AWS servers
+
+## What
+
+Proposes a way by which an SSH service might automatically discover and register AWS
+EC2 instances.
+
+## Why
+
+Currently when adding a new AWS server, it's required that Teleport be installed
+after the server has been provisioned which may be a slow process for organizations
+with large numbers of servers as it needs to be installed and then added to the
+teleport cluster
+
+With the changes described in this document, Teleport will be able to resolve the
+issues with adding AWS servers to Teleport clusters automatically.
+
+
+## Discovery
+
+Discovery can use a matcher similar to the `db_service/aws` matcher, however EC2
+instances will have an optional install command:
+
+```yaml
+ssh_service:
+  enabled: "yes"
+  aws:
+    invite_token: "iam-token-name"
+    matchers:
+    - types: ["ec2"]
+      regions: ["us-west-1"]
+      tags:
+        "teleport": "yes" # aws tags to match
+      ssm_command_document: ssm_command_document_name
+```
+
+The agent will use EC2's `DescribeInstances` API in order to list instances[1]. This
+will require the teleport SSH agent to include `ec2:DescribeInstances` as part of
+it's IAM permissions
+
+As with AWS database discover, new EC2 nodes will be discovered periodically on a 60
+second timer, as new nodes are found they will be added to the teleport cluster.
+
+In order to avoid attempting to reinstall teleport on top of an instance where it is
+already present the generated teleport config will include a static label indicating
+it was created via auto discovery.
+
+Example:
+```json
+{
+  "kind": "node",
+  "version": "v2",
+  "metadata": {
+    "name": "AWS_INSTANCE_ID",
+    "labels": {
+      "env": "example"
+      "teleport.dev/discovered-node": "yes"
+    },
+  },
+  "spec": {
+    "public_addr": "ec2-54-194-252-215.us-west-1.compute.amazonaws.com",
+    "hostname": "awsxyz"
+  }
+}
+```
+
+IAM join tokens will have a `tags` field added to the accounts so so only EC2
+instances with matching `tags` can be added using the token. the tags matching will
+default to `"*":"*"` to retain compatibility with previously generated IAM tokens
```

Comment:
Hmm... sorry, I just remembered the big downside of the EC2 join method is that it doesn't work on cloud specifically because it requires the `ec2:DescribeInstances` permissions on auth. Not great, sorry I didn't think this through more before commenting.

However, this might be a good opportunity to allow joining with the EC2 method through an SSH agent deployed on the customer's infra, I'm not sure how this would work yet but it could be doable. Would probably require an RFD of its own.